ethereum/eth/src/test/java/tech/pegasys/pantheon/ethereum/eth/sync/BlockPropagationManagerTest.java

```diff
@@ -561,4 +566,37 @@ public void shouldNotImportBlocksThatAreAlreadyBeingImported() {
 
     verify(ethScheduler, times(1)).scheduleSyncWorkerTask(any(Supplier.class));
   }
+
+  @Test
+  public void verifyBroadcastBlockInvocation() {
+    final BlockPropagationManager<Void> blockPropagationManager =
+        spy(
```

Comment:
I'd suggest using a mock here instead of a spy.  Spies are generally overly complicated and we don't need the real block propagation - just want to check that the broadcast method was called.

Reply:
why do you say they're overly complicated? if I change that to mock everything breaks in a confusing way

Reply:
something related to that generic component

Reply:
Because a spy is a combination of a mock and a real implementation.  So you're simultaneously saying this thing is within the scope of the unit I'm testing (thus using the real one) but also that it's outside of the unit (thus I want to check how it's interacted with).

And the voodoo required to make a spy work leads to a number of weird implications (e.g. that test you and I hit a few weeks back where moving a method reference to the constructor broke the tests because the spy no longer picked it up).

So for this test you don't want to spy on the `BlockPropagationManager`, you already have a mock `BlockBroadcaster` that you can assert on:

```


  @Test
  public void verifyBroadcastBlockInvocation() {
    blockchainUtil.importFirstBlocks(2);
    final Block block = blockchainUtil.getBlock(2);
    blockPropagationManager.start();

    // Setup peer and messages
    final RespondingEthPeer peer = EthProtocolManagerTestUtil.createPeer(ethProtocolManager, 0);

    final UInt256 totalDifficulty = fullBlockchain.getTotalDifficultyByHash(block.getHash()).get();
    final NewBlockMessage newBlockMessage = NewBlockMessage.create(block, totalDifficulty);

    // Broadcast message
    EthProtocolManagerTestUtil.broadcastMessage(ethProtocolManager, peer, newBlockMessage);

    final Responder responder = RespondingEthPeer.blockchainResponder(fullBlockchain);
    peer.respondWhile(responder, peer::hasOutstandingRequests);

    verify(blockBroadcaster, times(1)).propagate(block, totalDifficulty);
  }
```

And that test will fail because the total difficulty doesn't match.  The test is right, the code is wrong. :)  Total difficulty is being calculated as:
```
final UInt256 totalDifficulty =
          parent.getDifficulty().plus(block.getHeader().getDifficulty());
```
which is just parent difficulty + child difficulty when you need the sum of the difficulty for every block in the chain to get the total difficulty.  `final UInt256 totalDifficulty = protocolContext.getBlockchain().getTotalDifficultyByHash(parent.getHash()).get().plus(block.getHeader().getDifficulty());` will get you what you want.

Reply:
interesting. so that's just the difficulty of that one parent block in isolation, the parent block doesn't incorporate the difficulty of the blocks that came before it. && more interesting still -- the genesis block has an actual difficulty. anyway - thanks for pointing this out.